src/main/scala/com/databricks/labs/overwatch/utils/Structures.scala

```diff
@@ -194,47 +205,64 @@ private[overwatch] class ApiCallFailure(
+/**
+ * Specialized error to
+ *
+ * @param target          Pipeline Source throwing the error
+ * @param invalidReportDF details of keys causing the error
+ * @param fromCol         column name of the active start date or time
+ * @param untilCol        column name of the expiry date or time, null == active
+ */
+private[overwatch] class InvalidType2Input(
+                                            target: PipelineTable,
+                                            invalidReportDF: DataFrame,
+                                            fromCol: String,
+                                            untilCol: String
+                                          )
   extends Exception() with SparkSessionWrapper {
 
   import spark.implicits._
-  private val erroredRecordsReport = instanceDetails
-    .withColumn("API_Name", lower(trim('API_Name)))
+
+  private val erroredRecordsReport = target.asDF
     .join(
-      dfCheck
-        .select(
-          lower(trim('API_Name)).alias("API_Name"),
-          'rnk, 'rn, 'previousUntil,
-          datediff('activeFrom, 'previousUntil).alias("daysBetweenCurrentAndPrevious")
-        ),
-      Seq("API_Name")
+      invalidReportDF,
+      target.keys
     )
-    .orderBy('API_Name, 'activeFrom, 'activeUntil)
+    .orderBy(target.keys.map(col) ++ Array(col(fromCol), col(untilCol)): _*)
 
   println("InstanceDetails Error Report: ")
   erroredRecordsReport.show(numRows = 1000, false)
 
-  private val badRecords = dfCheck.count()
-  private val badRawKeys = dfCheck.select('API_Name).as[String].collect().mkString(", ")
-  private val errMsg = s"InstanceDetails Invalid: API_Name keys with errors are: $badRawKeys. " +
-    s"A total of $badRecords records were found in conflict. Each key (API_Name) must be unique for a given time period " +
-    s"(activeFrom --> activeUntil) AND the previous costs activeUntil must run through the previous date " +
-    s"such that the function 'datediff' returns 1. Please correct the instanceDetails table before continuing " +
-    s"with this module. "
+  private val badRecords = invalidReportDF.count()
+  private val badRawKeys = invalidReportDF.select(target.keys.map(col): _*).drop("organization_id").distinct()
+    .as[String].collect().mkString(", ")
+  private val errMsg = s"${target.tableFullName} Invalid: ${target.keys.mkString(", ")} keys with errors are: $badRawKeys. " +
+    s"A total of $badRecords records were found in conflict. " +
+    s"There cannot be any gaps or overlaps in time for a specific key such that " +
+    s"datediff($fromCol --> $untilCol) returns 0 for all keys. Please correct the data in the table " +
+    s"before continuing with Gold Pipeline."
   throw new BadConfigException(errMsg)
 }
 
-private[overwatch] class PipelineStateException(s: String, val target: Option[PipelineTable]) extends Exception(s) {}
+private[overwatch] class PipelineStateException(s: String, val target: Option[PipelineTable]) extends Exception(s) {
+  println(s)
```

Comment:
Remove `println` everywhere.

Reply:
why? without it the clean error message I've worked hard to construct doesn't appear in the STDOUT of the pipeline. These errors often don't hard-fail the pipeline but rather soft-fail the module...so yeah, they'd be in the log4j but the stdout is much friendlier. I want users to know when they're getting an error, especially in a notebook because that's primarily when they are setting up their first deployment.